### PR TITLE
#125 Select checkboxes can scroll to match when using input

### PR DESCRIPTION
**Pull request by veryprofessionaldodo**
| -| - | 
| ---| --- | 
| Issue | https://github.com/ripe-tech/ripe-util-vue/issues/125 |
| Dependencies | None |
| Decisions | Select checkboxes didn't have the feature to scroll to the closest match when input is given, so this PR adds that feature.  |
| Animated GIF | Peek 2021-07-22 09-31 |

### CODE REVIEW

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue [hunk 1]
```diff
@@ -1,21 +1,18 @@
...
-            ...selectProps
-        }"
-    >
-        <template v-slot:checkbox-group>
-            <div class="checkboxes" v-on:click.stop>
-                <checkbox-group v-bind:items="_items" v-bind:values.sync="valuesData" />
-            </div>
-        </template>
-    </select-ripe>
+    <div v-on:keydown.exact.stop="event => onKey(event.key)">
```
Comment (NFSS10) [inline vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:2]: This seems a very big change that has everything to break the select-checboxes in other places. I think this can't be done this way.

If the reason you added this div was just to do get the `on:key` event, try to just emit [this event in select-ripe](https://github.com/ripe-tech/ripe-components-vue/blob/a2b9d2aa0e199d59211232f6cf15d950ef052206/vue/components/ui/molecules/select/select.vue#L26),

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue [hunk 2]
```diff
@@ -170,6 +177,45 @@ export const SelectCheckboxes = {
             },
             deep: true
         }
+    },
+    methods: {
+        onKey(key) {
```
Comment (NFSS10) [inline vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:182]: This seems to be [select-ripe `onKey` logic ](https://github.com/ripe-tech/ripe-components-vue/blob/a2b9d2aa0e199d59211232f6cf15d950ef052206/vue/components/ui/molecules/select/select.vue#L369) I would call the method something like `keyDebounce` and add this to a mixin (same for other logic you think it can be used in a general context), and share between the two components, or probably better yet, just have the select-ripe sending the key event and the `keyBuffer` with the `on:key` event and use that

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue [hunk 3]
```diff
@@ -170,6 +177,45 @@ export const SelectCheckboxes = {
...
+            this.keyBuffer += key.toUpperCase();
+            this.keyTimestamp = Date.now();
+            this.scrollToBestOption();
+        },
+        scrollToBestOption() {
+            const index = this._items.findIndex(option =>
+                option.label?.toUpperCase().startsWith(this.keyBuffer)
+            );
+
+            if (index === null || index < 0 || index >= this._items.length) return;
```
Comment (NFSS10) [inline vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:196]: `if (index === -1) return;`

The return value is -1 if not found or >= 0, you shouldn't have all of this conditions, see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex#return_value

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue [hunk 4]
```diff
@@ -117,10 +114,20 @@ export const SelectCheckboxes = {
...
+         * Amount of time before current keybuffer is discarded.
+         */
+        keyTimeout: {
+            type: Number,
+            default: 500
         }
     },
     data: function() {
         return {
+            highlighted: null,
```
Comment (NFSS10) [inline vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:128]: dead variable

**Review by NFSS10 — CHANGES_REQUESTED** [review]

**Comment by NFSS10** [comment]
@veryprofessionaldodo also, the build is failing

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue (hunk 1 above)
Comment (veryprofessionaldodo) [reply vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:2]: Emitting an event in the select-ripe also worked, I did test that. I see what you mean with that, and I'll do those changes!

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue (hunk 3 above)
Comment (veryprofessionaldodo) [reply vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:196]: Makes total sense, nice catch!

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue (hunk 2 above)
Comment (veryprofessionaldodo) [reply vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:182]: That makes total sense, adding that in!

File: vue/components/ui/molecules/select/select.vue [hunk 5]
```diff
@@ -373,6 +357,7 @@ export const Select = {
             }
             this.keyBuffer += key.toUpperCase();
             this.keyTimestamp = Date.now();
+            this.$emit("select:keydown", this.keyBuffer);
```
Comment (NFSS10) [inline vue/components/ui/molecules/select/select.vue:360]: `this.$emit("select:keydown", key, this.keyBuffer);`

You should also emit the event and the key, As this `onKey` handler doesn't receive the `event` we are only emitting `key` :/

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue [hunk 6]
```diff
@@ -3,15 +3,11 @@
...
-        v-bind="{
-            autoScroll: false,
-            maxHeight: 210,
-            ...selectProps
-        }"
+        ref="select"
+        v-on:select:keydown="scrollToInput"
     >
         <template v-slot:checkbox-group>
-            <div class="checkboxes" v-on:click.stop>
```
Comment (NFSS10) [inline vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:14]: Why did you remove this? This can break things in other places.

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue [hunk 7]
```diff
@@ -3,15 +3,11 @@
         class="select-checkboxes"
         v-bind:options="selectOptions"
         v-bind:value="'checkbox-group'"
-        v-bind:disabled="disabled"
-        v-bind="{
-            autoScroll: false,
-            maxHeight: 210,
-            ...selectProps
-        }"
```
Comment (NFSS10) [inline vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:11]: Why did you remove this? This can break things in other places.

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue [hunk 8]
```diff
@@ -170,6 +169,21 @@ export const SelectCheckboxes = {
             },
             deep: true
         }
+    },
+    methods: {
+        scrollToInput(keyBuffer) {
```
Comment (NFSS10) [inline vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:174]: `scrollToCheckbox`?

File: vue/mixins/utils.js [hunk 9]
```diff
@@ -26,6 +26,24 @@ export const utilsMixin = {
...
+            }
+
+            const indexHeight = elements[index].offsetHeight;
+            const indexEnd = indexStart + indexHeight;
+
+            if (indexStart < visibleStart) {
+                dropdown.scrollTop = indexStart;
+            } else if (indexEnd > visibleEnd) {
+                dropdown.scrollTop = indexEnd - dropdown.clientHeight;
+            }
```
Comment (NFSS10) [inline vue/mixins/utils.js:46]: We can probably name this to be more generic. This should work with any tipe of container and not just with dropdowns right?

Comment (NFSS10) [reply vue/mixins/utils.js:46]: Also as we are making this a generic util method it's probably better to jsdoc it so other people can easily know what it does

**Review by NFSS10 — CHANGES_REQUESTED** [review]

File: vue/mixins/utils.js (hunk 9 above)
Comment (veryprofessionaldodo) [reply vue/mixins/utils.js:46]: Fair points, I'll do both!

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue (hunk 8 above)
Comment (veryprofessionaldodo) [reply vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:174]: Yeah, I didn't like the name I had either, yours is more appropriate.

File: vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue (hunk 6 above)
Comment (veryprofessionaldodo) [reply vue/components/ui/molecules/select-checkboxes/select-checkboxes.vue:14]: It was meant to test things, as I was having difficulties getting the input. Forgot to put it back, similar to the `bind` and `disabled` parameters. Already re-added them, thanks!

**Review by NFSS10 — CHANGES_REQUESTED** [review]

File: vue/mixins/utils.js [hunk 10]
```diff
@@ -26,6 +26,31 @@ export const utilsMixin = {
...
         buildSlug(value) {
             return buildSlug(value);
+        },
+        /**
+         * Scrolls to the right element for a container,
+         * given the index of the element to scroll to.
+         * @param {*} container The parent container;
+         * @param {*} elements List of elements to iterate over;
+         * @param {*} index The index of the target element;
+         */
```
Comment (gcandal) [inline vue/mixins/utils.js:36]: Newline + types
```suggestion
        /**
         * Scrolls to the right element for a container,
         * given the index of the element to scroll to.
         *
         * @param {HTMLElement} container The parent container;
         * @param {HTMLCollection} elements List of elements to iterate over;
         * @param {Number} index The index of the target element;
         */

```

**Review by gcandal — APPROVED** [review]

**Review by joamag — APPROVED** [review]

**Comment by joamag** [comment]
@veryprofessionaldodo good job... please confirm that the solution still works after the refactor

**Comment by veryprofessionaldodo** [comment]
Can confirm, solution is working as expected using the latest commits from master! :)
